File: ios/Impermanence/Impermanence/Models/Bell.swift
//
//  Bell.swift
//  Impermanence
//
//  Created by Alex Ellis on 8/5/23.
//

import Foundation

struct Bell: Codable, Equatable {
    var soundId: Int
    var numRings: Int

    static let minRings = 1
    static let maxRings = 12

    var sound: BellSound {
        BellCatalog.sound(for: soundId)
    }
}

extension Bell {
    static let singleBell: Bell = Bell(soundId: 0, numRings: 1)
    static let repeatedBell: Bell = Bell(soundId: 0, numRings: 3)

    static func bell(for sound: BellSound, rings: Int) -> Bell {
        Bell(soundId: sound.id, numRings: rings).sanitized()
    }

    func sanitized() -> Bell {
        let validSoundId = BellCatalog.sounds.contains(where: { $0.id == soundId }) ? soundId : Bell.singleBell.soundId
        let validRings = min(max(numRings, Bell.minRings), Bell.maxRings)
        return Bell(soundId: validSoundId, numRings: validRings)
    }
}

File: android/app/src/main/java/com/impermanence/impermanence/model/Bell.kt
package com.impermanence.impermanence.model

import kotlinx.serialization.Serializable

@Serializable
data class Bell(
    val soundId: Int,
    val numRings: Int
) {
    val sound: BellSound
        get() = BellCatalog.soundFor(soundId)

    companion object {
        val SingleBell = Bell(soundId = 0, numRings = 1)
        val RepeatedBell = Bell(soundId = 0, numRings = 3)

        fun bellFor(sound: BellSound, rings: Int): Bell = Bell(soundId = sound.id, numRings = rings)
    }
}
